runners/direct-java/src/main/java/org/apache/beam/runners/direct/CommittedResult.java

```diff
@@ -49,12 +50,23 @@
    */
   public abstract Iterable<? extends CommittedBundle<?>> getOutputs();
 
+  /**
+   * Returns if the transform that produced this result produced outputs.
+   *
+   * <p>Transforms that produce output via modifying the state of the runner (e.g.
+   * {@link CreatePCollectionView}) should explicitly set this to true. If {@link #getOutputs()}
+   * returns a nonempty iterable, this will also return true.
+   */
+  public abstract boolean producedOutputs();
```

Comment:
This seems like a good candidate for the mantra that booleans rarely stay boolean. In particular, you already have three cases:

1. Produced elements
2. Wrote to a side input area (which is probably subject to change as the runner evolves, and just means "created some work somewhere that isn't new elements)
3. Did nothing that could unblock new work

So maybe a little enum with a method that answers this boolean question.